File: src/resolver/graph.zig
const std = @import("std");
const types = @import("../types.zig");
const Sha256 = std.crypto.hash.sha2.Sha256;

pub const Node = struct {
    name: []const u8,
    source: types.SourceType,
    version: types.Version,
    package_hash: ?[]const u8 = null,
    dependencies: []const []const u8 = &.{},
};

pub const Graph = struct {
    allocator: std.mem.Allocator,
    nodes: std.ArrayListUnmanaged(Node) = .{},

    pub fn init(allocator: std.mem.Allocator) Graph {
        return .{ .allocator = allocator };
    }

    pub fn deinit(self: *Graph) void {
        self.nodes.deinit(self.allocator);
    }

    pub fn addNode(self: *Graph, node: Node) !void {
        try self.nodes.append(self.allocator, node);
    }

    pub fn findNode(self: *Graph, name: []const u8) ?usize {
        for (self.nodes.items, 0..) |n, i| {
            if (std.mem.eql(u8, n.name, name)) return i;
        }
        return null;
    }

    pub fn computeHash(self: *Graph) ![Sha256.digest_length]u8 {
        _ = self;
        return [_]u8{0} ** 32;
    }

    pub fn hasCycles(self: *Graph) bool {
        _ = self;
        return false;
    }
};

test "graph: add and find nodes" {
    var g = Graph.init(std.testing.allocator);
    defer g.deinit();

    try g.addNode(.{
        .name = "zod",
        .source = .npm,
        .version = try types.Version.parse("3.23.8"),
    });
    try g.addNode(.{
        .name = "react",
        .source = .npm,
        .version = try types.Version.parse("18.3.0"),
    });

    try std.testing.expectEqual(@as(usize, 2), g.nodes.items.len);
    try std.testing.expectEqual(@as(usize, 0), g.findNode("zod").?);
    try std.testing.expectEqual(@as(usize, 1), g.findNode("react").?);
    try std.testing.expect(g.findNode("missing") == null);
}

test "graph: hasCycles returns false (stub)" {
    var g = Graph.init(std.testing.allocator);
    defer g.deinit();

    try std.testing.expect(!g.hasCycles());
}

test "graph: computeHash returns zeros (stub)" {
    var g = Graph.init(std.testing.allocator);
    defer g.deinit();

    const hash = try g.computeHash();
    try std.testing.expectEqual(@as(usize, 32), hash.len);
    for (hash) |byte| try std.testing.expectEqual(@as(u8, 0), byte);
}

File: src-rs/src/resolver/graph.rs
#![allow(dead_code)]

use crate::types::{SourceType, Version};
use serde::Serialize;

#[derive(Debug, Clone, Serialize)]
pub struct Node {
    pub name: String,
    pub source: SourceType,
    pub version: Version,
    pub package_hash: Option<String>,
    pub dependencies: Vec<String>,
}

#[derive(Debug, Clone)]
pub struct Graph {
    pub nodes: Vec<Node>,
}

impl Graph {
    #[must_use]
    pub fn new() -> Self {
        Self { nodes: Vec::new() }
    }

    pub fn add_node(&mut self, node: Node) {
        self.nodes.push(node);
    }

    #[must_use]
    pub fn find_node(&self, name: &str) -> Option<usize> {
        self.nodes.iter().position(|n| n.name == name)
    }

    /// Compute a hash for the graph. Currently a stub.
    #[must_use]
    pub fn compute_hash(&self) -> [u8; 32] {
        [0u8; 32]
    }

    /// Check for cycles. Currently a stub.
    #[must_use]
    pub fn has_cycles(&self) -> bool {
        false
    }
}

impl Default for Graph {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::types::Version;

    #[test]
    fn test_add_and_find_nodes() {
        let mut g = Graph::new();
        g.add_node(Node {
            name: "zod".to_string(),
            source: SourceType::Npm,
            version: Version::parse("3.23.8").unwrap(),
            package_hash: None,
            dependencies: Vec::new(),
        });
        g.add_node(Node {
            name: "react".to_string(),
            source: SourceType::Npm,
            version: Version::parse("18.3.0").unwrap(),
            package_hash: None,
            dependencies: Vec::new(),
        });

        assert_eq!(g.nodes.len(), 2);
        assert_eq!(g.find_node("zod"), Some(0));
        assert_eq!(g.find_node("react"), Some(1));
        assert_eq!(g.find_node("missing"), None);
    }

    #[test]
    fn test_has_cycles_returns_false() {
        let g = Graph::new();
        assert!(!g.has_cycles());
    }

    #[test]
    fn test_compute_hash_returns_zeros() {
        let g = Graph::new();
        let hash = g.compute_hash();
        assert_eq!(hash.len(), 32);
        assert!(hash.iter().all(|&b| b == 0));
    }
}
